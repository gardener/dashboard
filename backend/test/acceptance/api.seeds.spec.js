//
// SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

'use strict'

const { keyBy, pick } = require('lodash')
const { mockRequest } = require('@gardener-dashboard/request')

describe('api', function () {
  let agent

  beforeAll(() => {
    agent = createAgent()
  })

  afterAll(() => {
    return agent.close()
  })

  beforeEach(() => {
    mockRequest.mockReset()
  })

  describe('seeds', function () {
    const id = 'john.doe@example.org'
    const user = fixtures.auth.createUser({ id })

    it('should return all seeds', async function () {
      mockRequest.mockImplementationOnce(fixtures.auth.reviewSelfSubjectAccess())

      const res = await agent
        .get('/api/seeds')
        .set('cookie', await user.cookie)
        .expect('content-type', /json/)
        .expect(200)

      expect(mockRequest).toBeCalledTimes(1)
      expect(mockRequest.mock.calls[0]).toEqual([
        {
          ...pick(fixtures.kube, [':scheme', ':authority']),
          authorization: `Bearer ${await user.bearer}`,
          ':method': 'post',
          ':path': '/apis/authorization.k8s.io/v1/selfsubjectaccessreviews'
        },
        {
          apiVersion: 'authorization.k8s.io/v1',
          kind: 'SelfSubjectAccessReview',
          spec: expect.objectContaining({
            resourceAttributes: {
              group: 'core.gardener.cloud',
              name: undefined,
              resource: 'seeds',
              verb: 'list'
            }
          })
        }
      ])

      expect(res.body).toHaveLength(8)
      const seeds = keyBy(res.body, 'metadata.name')
      expect(seeds['infra1-seed'].metadata.unreachable).toBe(false)
      expect(seeds['infra3-seed'].metadata.unreachable).toBe(true)
    })
  })
})
