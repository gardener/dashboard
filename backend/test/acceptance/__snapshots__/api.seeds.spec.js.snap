// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api seeds should return all seeds 1`] = `
Array [
  Array [
    Object {
      ":authority": "kubernetes:6443",
      ":method": "post",
      ":path": "/apis/authorization.k8s.io/v1/selfsubjectaccessreviews",
      ":scheme": "https",
      "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImpvaG4uZG9lQGV4YW1wbGUub3JnIiwiaWF0IjoxNTc3ODM2ODAwLCJhdWQiOlsiZ2FyZGVuZXIiXSwiZXhwIjozMTU1NzE2ODAwLCJqdGkiOiJqdGkifQ.LkQ9PEN893UNTsZZn2Ux_CAYNOoQ2ISboWuHiAc5HHU",
    },
    Object {
      "apiVersion": "authorization.k8s.io/v1",
      "kind": "SelfSubjectAccessReview",
      "spec": Object {
        "nonResourceAttributes": undefined,
        "resourceAttributes": Object {
          "group": "core.gardener.cloud",
          "resource": "seeds",
          "verb": "list",
        },
      },
    },
  ],
]
`;

exports[`api seeds should return all seeds 2`] = `
Array [
  Object {
    "data": Object {
      "region": "foo-east",
      "type": "infra1",
      "unprotected": false,
      "visible": false,
    },
    "metadata": Object {
      "name": "soil-infra1",
      "unreachable": false,
    },
  },
  Object {
    "data": Object {
      "region": "foo-east",
      "type": "infra1",
      "unprotected": true,
      "visible": true,
    },
    "metadata": Object {
      "name": "infra1-seed",
      "unreachable": false,
    },
  },
  Object {
    "data": Object {
      "region": "foo-west",
      "type": "infra1",
      "unprotected": true,
      "visible": true,
    },
    "metadata": Object {
      "name": "infra1-seed2",
      "unreachable": false,
    },
  },
  Object {
    "data": Object {
      "region": "foo-europe",
      "type": "infra3",
      "unprotected": true,
      "visible": true,
    },
    "metadata": Object {
      "name": "infra3-seed",
      "unreachable": true,
    },
  },
  Object {
    "data": Object {
      "region": "foo-south",
      "type": "infra1",
      "unprotected": true,
      "visible": true,
    },
    "metadata": Object {
      "name": "infra4-seed-without-secretRef",
      "unreachable": false,
    },
  },
  Object {
    "data": Object {
      "region": "foo-europe",
      "type": "infra3",
      "unprotected": true,
      "visible": true,
    },
    "metadata": Object {
      "name": "infra3-seed-with-selector",
      "unreachable": false,
    },
  },
  Object {
    "data": Object {
      "region": "foo-europe",
      "type": "infra3",
      "unprotected": false,
      "visible": true,
    },
    "metadata": Object {
      "name": "infra3-seed-protected",
      "unreachable": false,
    },
  },
  Object {
    "data": Object {
      "region": "foo-europe",
      "type": "infra3",
      "unprotected": true,
      "visible": false,
    },
    "metadata": Object {
      "name": "infra3-seed-invisible",
      "unreachable": false,
    },
  },
]
`;
